lib/r10k/puppetfile.rb

```diff
@@ -144,7 +162,9 @@ def add_module(name, args)
     @managed_content[install_path] = Array.new unless @managed_content.has_key?(install_path)
     @managed_content[install_path] << mod.name
 
-    @modules << mod
+    # We want to manage the content (and not purge it) even if we do not want
+    # to sync modules that we assume unchanged.
+    no_change ? @modules : @modules << mod
```

Comment:
I worry about focusing effort on the Puppetfile, when I see the future direction being more "modules attached to an environment", with the Puppetfile being kind of a legacy way to attach them.

Is it possible to take something like this approach instead?

1. Add something like `#state_change?` and `#state_change=` methods to module objects.
2. In the Puppetfile, set `mod.state_change = true/false/:unknown` as appropriate.
3. Instead of adding logic to mask the existence of the module here in the Puppetfile code, figure out where the module is actually synced/verified, and put an `... unless !mod.state_change?` there. Is that [here|https://github.com/puppetlabs/r10k/blob/aab476ab05432c2a7e5361cd68524e19a53b9f74/lib/r10k/environment/with_modules.rb#L80-L82]? (no idea)

That approach would also preserve the ability to "see" all modules attached to an environment, regardless of what you're planning on doing to/with them.

Again, mostly I'm just worried about embedding important logic in the Puppetfile specifically, given that the Puppetfile is only one way to define environment content. At some point a general encapsulation/cleanup of how environments and modules relate would be nice, but in the meantime I think not implementing new stuff as Puppetfile-specific would be a good idea.